### PR TITLE
Merge flattened css into main branch

### PR DESCRIPTION
**Pull request by jacoblogan**
#### Description of changes
This pull request our current styles across react, angular, and vue to use flattened css selectors following the BEM styling syntax.  This is to produce a better experience for customers who are using css frameworks and tools.  By lowering our selector specificity it allows our customers to in turn use lower specificity selectors to overwrite our default styles.

- [x] PR description included
- [x] `yarn test` passes

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by zchenwei — APPROVED** [review]
LGTM!

File: packages/angular/projects/ui-angular/src/lib/primitives/button/button.component.ts [hunk 1]
```diff
@@ -15,9 +15,22 @@ export class ButtonComponent implements OnInit {
   @HostBinding('attr.data-fullwidth') fullWidthAttr: boolean | string;
   @HostBinding('attr.data-size') sizeAttr: string;
   @HostBinding('attr.data-variation') variationAttr: string;
-  @HostBinding('class.amplify-button') defaultClass = true;
   @HostBinding('style.font-weight') fontWeightAttr: string;
 
+  @HostBinding('class') get classNames() {
+    let className = 'amplify-button';
+    if(this.variation){
```
Comment (reesscot) [inline packages/angular/projects/ui-angular/src/lib/primitives/button/button.component.ts:22]: nit: spacing
```suggestion
    if (this.variation){
```

File: packages/angular/projects/ui-angular/src/lib/primitives/button/button.component.ts [hunk 2]
```diff
@@ -15,9 +15,22 @@ export class ButtonComponent implements OnInit {
...
+      className += ` amplify-button--${this.variation}`;
+    }
+    if(this.size){
```
Comment (reesscot) [inline packages/angular/projects/ui-angular/src/lib/primitives/button/button.component.ts:25]:
```suggestion
    if (this.size){
```

File: packages/react/src/primitives/shared/utils.ts [hunk 3]
```diff
@@ -159,3 +161,34 @@ export const mergeVariantsAndOverrides = (
     ...merged,
   };
 };
+
+type Modifiers = string | number | null;
+/**
```
Comment (reesscot) [inline packages/react/src/primitives/shared/utils.ts:166]:
```suggestion

/**
```

File: packages/react/src/primitives/shared/utils.ts [hunk 4]
```diff
@@ -159,3 +161,35 @@ export const mergeVariantsAndOverrides = (
...
+type Modifiers = string | number | null;
+
+/**
+ * This helper function creates modifier class names that are used for our flat BEM styling
+ * it takes in a base and modifier and returns the modified class if a modifier was passed in and null otherwise
+ * @param base The base class of the output
+ * @param modifier The modifier to add onto the base
+ * @returns the modified class name or null
+ */
+export const classNameModifier = (
```
Comment (reesscot) [inline packages/react/src/primitives/shared/utils.ts:174]: Love these utils!

File: packages/react/src/primitives/Input/Input.tsx [hunk 5]
```diff
@@ -33,40 +34,47 @@ const InputPrimitive: Primitive<InputProps, 'input'> = (
...
-    type={type}
-    value={value}
-    {...rest}
-  />
-);
+) => {
+  const componentClasses = classNames(
+    ComponentClassNames.Input,
+    ComponentClassNames.FieldGroupControl,
+    classNameModifier(ComponentClassNames.Input, variation),
```
Comment (reesscot) [inline packages/react/src/primitives/Input/Input.tsx:41]: Need to update unit tests to reflect the new classnames being added to each primitive

File: packages/react/src/primitives/View/View.tsx [hunk 6]
```diff
@@ -27,6 +27,7 @@ const ViewPrimitive = <Element extends ElementType = 'div'>(
       'aria-label': ariaLabel,
       'data-testid': testId,
       disabled: isDisabled,
+      isDisabled: isDisabled,
```
Comment (reesscot) [inline packages/react/src/primitives/View/View.tsx:30]: Why is this needed?

Comment (jacoblogan) [reply packages/react/src/primitives/View/View.tsx:30]: this was added in because of the `FieldGroupIconButton` which is using the `as` property to render our primitive `Button` but the `isDisabled` flag was not being propagated down to the Button primitive.

**Review by joebuono — APPROVED** [review]

**Review by wlee221 — APPROVED** [review]

File: packages/angular/projects/ui-angular/src/lib/components/authenticator/components/confirm-sign-in/confirm-sign-in.component.html [hunk 7]
```diff
@@ -6,7 +6,7 @@
     [disabled]="authenticator.isPending"
   >
     <amplify-slot name="confirm-sign-in-header" [context]="context">
-      <h3 class="amplify-heading">{{ headerText }}</h3>
+      <h3 class="amplify-heading amplify-heading--3">{{ headerText }}</h3>
```
[the remark this answers is not in the document]
Comment (jacoblogan) [reply packages/angular/projects/ui-angular/src/lib/components/authenticator/components/confirm-sign-in/confirm-sign-in.component.html:9]: yes Vue authenticator needs this class as well, but instead of adding it manually like we are doing here in angular, Vue has a basic Heading component defined that it is using and so the heading class is being added there much like it is in React.
